Source/ACE.Server/Factories/LootGenerationFactory_Melee.cs

```diff
@@ -33,37 +33,47 @@ public static WorldObject CreateMeleeWeapon(TreasureDeath profile, bool isMagica
             {
                 case 0:
                     // Heavy Weapons
-                    subtype = ThreadSafeRandom.Next(0, 22);
+                    subtype = ThreadSafeRandom.Next(0, LootTables.HeavyWeaponsMatrix.Length - 1);
                     weaponWeenie = LootTables.HeavyWeaponsMatrix[subtype][eleType];
                     break;
                 case 1:
                     // Light Weapons
-                    subtype = ThreadSafeRandom.Next(0, 19);
+                    subtype = ThreadSafeRandom.Next(0, LootTables.LightWeaponsMatrix.Length - 1);
                     weaponWeenie = LootTables.LightWeaponsMatrix[subtype][eleType];
                     break;
                 case 2:
                     // Finesse Weapons;
-                    subtype = ThreadSafeRandom.Next(0, 22);
+                    subtype = ThreadSafeRandom.Next(0, LootTables.FinesseWeaponsMatrix.Length - 1);
                     weaponWeenie = LootTables.FinesseWeaponsMatrix[subtype][eleType];
                     break;
                 default:
                     // Two handed
-                    subtype = ThreadSafeRandom.Next(0, 11);
+                    subtype = ThreadSafeRandom.Next(0, LootTables.TwoHandedWeaponsMatrix.Length - 1);
                     weaponWeenie = LootTables.TwoHandedWeaponsMatrix[subtype][eleType];
                     break;
             }
 
             var wo = WorldObjectFactory.CreateNewWorldObject((uint)weaponWeenie);
 
             if (wo != null && mutate)
-                MutateMeleeWeapon(wo, profile, isMagical, weaponType, subtype);
+                if (!MutateMeleeWeapon(wo, profile, isMagical))
+                    wo = null;
 
             return wo;
         }
 
-        private static void MutateMeleeWeapon(WorldObject wo, TreasureDeath profile, bool isMagical, int weaponType, int subtype)
+        private static bool MutateMeleeWeapon(WorldObject wo, TreasureDeath profile, bool isMagical)
         {
-            Skill wieldSkillType = Skill.None;
+            if ((wo.WeaponSkill != Skill.HeavyWeapons || wo.WeaponSkill != Skill.LightWeapons || wo.WeaponSkill != Skill.FinesseWeapons || wo.WeaponSkill != Skill.TwoHandedCombat)
```

Comment:
This is written incorrectly, and will cause every mutation of every melee weapon to fail. The intended logic here is &&, not ||

This should also be written as a switch/case statement, as it would execute faster than tons of separate if checks